### PR TITLE
[Simulator] Further improvements (reload radio, momentary switches, etc.)

### LINKED ISSUE #1627: Temporary hold SH switch down in Companion (Taranis)
In Companion, it would be nice if it were possible to temporary hold SH switch at its down position (maybe in a similar way as the Hold X and Hold Y boxes of the sticks) in order to simulate sequences like: SH-down / do something  / SH-up.

### PR DESCRIPTION
**Pull request by mpaperno**
* Add ability to reload all radio data w/out exiting simulator (will also adjust UI for any hardware changes) (helps with #4159);
 * Only update channel/logical switch/GVar output displays if the corresponding tab is actually visible (further CPU usage reductions);
 * Implement momentary toggle switches with "latch" function (closes #1627);
 * Using mouse wheel over multi-position knobs now properly rotates them to detent positions.
 * Fix throttle stick position/lock not being set properly upon repeated launches from Companion.

This is probably the last button that will fit up on the top toolbar... will need to switch to icons-only and/or a menu for further additions.

image

### CODE REVIEW

**Comment by 3djc** [comment]
Won't compile with the QT 5.3 of building machine:
In file included from /home/opentx/opentx/companion/src/simulation/simulatordialog.cpp:28:0:
/home/opentx/opentx/companion/src/simulation/widgets/radioswitchwidget.h: In member function ‘void RadioSwitchWidget::onValueChanged(int)’:
/home/opentx/opentx/companion/src/simulation/widgets/radioswitchwidget.h:121:77: error: no matching function for call to ‘QTimer::singleShot(int, RadioSwitchWidget*, void (RadioSwitchWidget::*)())’
         QTimer::singleShot(500, this, &RadioSwitchWidget::onMomentaryTimeout);

**Comment by projectkk2glider** [comment]
Testing this, works awesome.

One wish: momentary 2 pos switch returns to OFF position even if I hold it on the ON position with the mouse. I have setups that require SH to be ON for more time that is predefined in simulator. 

Therefore I would prefer if SH stays ON as long as I hold it there with the mouse. Is that possible?

**Comment by kilrah** [comment]
Last step will be an X9E skin, @mhotar? ;)

**Comment by mpaperno** [comment]
Damjan, great point.  I've just been flicking the switches mostly with mouse wheel and hadn't thought of "holding it down" issue.  So now it works both ways, much better I think.

And yes, the X9E seems a bit left out...  There's even some annoying conditional code in the X9D skin to handle the (slight) differences in controls.  ;-)

**Comment by projectkk2glider** [comment]
Momentary now works perfect :+1:

File: companion/src/simulation/widgets/lcdwidget.h [hunk 1]
```diff
@@ -72,9 +72,7 @@ class LcdWidget : public QWidget
 
     void setBackgroundColor(QColor color)
```
Comment (bsongis) [inline companion/src/simulation/widgets/lcdwidget.h]: Perhap it should be `const QColor &`

**Comment by bsongis** [comment]
Thanks!

File: companion/src/simulation/widgets/lcdwidget.h (hunk 1 above)
Comment (mpaperno) [inline companion/src/simulation/widgets/lcdwidget.h]: Yes it should!  Fixed in next version  :)  Thanks!

**Pull request merged**